### PR TITLE
CSS-3454: Backup and restore integration test

### PR DESCRIPTION
**Pull request by kelkawi-a**
This PR addresses charms' [Definition of Great Stage 2](https://docs.google.com/document/d/1aB7YrnVEqXWgzhZXQf6gOWZnbV9THjvTrW0KlCuLUpg/edit#) "Backup and Restore" goal. By default, the Temporal server and its services are stateless and rely on the database to store workflow execution status. To confirm this, I have done the following:

- Written an integration test which starts a running workflow that accepts signals, and can only complete execution when receiving an exit signal. After submitting a few signals, we simulate a charm crash by removing the application completely and re-deploying it, and continue the workflow execution after the application is re-deployed.
- Resolved a small bug which did not send the Temporal charm into a blocked state when the db relation was broken because of the way we were updating the database connections in the charm's state.
- Some light refactoring.

### CODE REVIEW

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -33,3 +31,12 @@ async def test_ui_relation(self, ops_test: OpsTest):
...
         await run_sample_workflow(ops_test)
+
+    async def test_charm_crash(self, ops_test: OpsTest):
+        """Test backup and restore functionality.
+
+        This tests the charm's ability to continue workflow execution after simulating
+        a crash in the charm. Essentially, it should prove that the charm is stateless
+        and relies only on the db to store its workflow execution status.
+        """
+        await run_signal_workflow(ops_test)
```
Comment (TeodorPt) [inline tests/integration/test_charm.py:42]: why is this in the helper and not directly here in the test?

**Review by TeodorPt — APPROVED** [review]
LGTM!

File: tests/integration/temporal_client/workflows.py [hunk 2]
```diff
@@ -28,3 +30,48 @@ async def run(self, name: str) -> str:
...
+        """Workflow execution method.
+
+        Returns:
+            Workflow execution result.
+        """
+        # Continually handle from queue or wait for exit to be received
+        greetings: List[str] = []
+        while True:
+            # Wait for queue item or exit
+            await workflow.wait_condition(lambda: not self._pending_greetings.empty() or self._exit)
```
Comment (kian99) [inline tests/integration/temporal_client/workflows.py:55]: Out of curosity, how does it wait under-the-hood for the variable change? If you're aware.

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -33,3 +41,57 @@ async def test_ui_relation(self, ops_test: OpsTest):
...
         await run_sample_workflow(ops_test)
+
+    async def test_charm_crash(self, ops_test: OpsTest):
+        """Test backup and restore functionality.
+
+        This tests the charm's ability to continue workflow execution after simulating
+        a crash in the charm. Essentially, it should prove that the charm is stateless
+        and relies only on the db to store its workflow execution status.
+        """
+        # await run_signal_workflow(ops_test)
```
Comment (kian99) [inline tests/integration/test_charm.py:52]: Commented out line? Should that be added or completely removed.

**Review by kian99 — APPROVED** [review]

File: tests/integration/temporal_client/workflows.py (hunk 2 above)
Comment (kelkawi-a) [reply tests/integration/temporal_client/workflows.py:55]: This is the implementation I found in their [SDK](https://github.com/temporalio/sdk-python/blob/a4224a20189c8c8078edbc66482c31860193c1b3/temporalio/workflow.py#L709), hope it helps!

**Pull request merged**